### PR TITLE
Complete rework

### PR DESCRIPTION
**Pull request by Rafi594**
## Problem
- *package is broken*

## Solution
- *apply example_ynh*
- *fix several issues*
- *Upgrade to version 5.1.0*

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Package_check results
---
[Build Status/badge/icon)](http://ci-apps-dev.yunohost.org/jenkins/job/minetest_ynh%20PR10%20(Rafi59)/)

### CODE REVIEW

**Comment by vincentux** [comment]
Install Minetest 5.0 fail :  
Voici les logs  
https://framabin.org/p/?bf0380c7ab2ad5b2#1J2yyd80mk+OOsZ3UeCVMa1q9Xn7Omh3e7CH2ffgc7g=

**Comment by vincentux** [comment]
Alors j'ai essayé d'installer la version 0.4.* et ensuite de faire une mise à jour avec ton dépôt,  

```
sudo yunohost app upgrade minetest -u https://github.com/Rafi594/minetest_ynh
Info: Now upgrading app minetest…
Success! minetest has been upgraded
Success! The SSOwat configuration has been generated
Success! Upgrade complete
```
Mais quand je lance la version client (5.0.1) chez moi pour me connecter sur le serveur, j'ai un message d'avertissement comme quoi les versions sont différente...  
Comment vérifier la version installé sur Yunohost ?

**Comment by Rafi594** [comment]
I'll fix the Postgresql option

**Comment by Rafi594** [comment]
Anyone on @YunoHost-Apps/apps-group can review this and give me access to the repository ?

File: scripts/upgrade [hunk 1]
```diff
@@ -14,28 +12,131 @@ source /usr/share/yunohost/helpers
...
+# Create a dedicated systemd config
+ynh_add_systemd_config
+
+#=================================================
+# SPECIFIC UPGRADE
+#=================================================
+# Compile Minetest
+cd /opt/yunohost/$app
+# Download Minetest Game
+git clone --depth 1 https://github.com/minetest/minetest_game.git games/minetest_game -b stable-5
```
Comment (kay0u) [inline scripts/upgrade]: Same here, use:
`ynh_setup_source --dest_dir=/opt/yunohost/$app/games/minetest_game --source_id=minetest_game`

File: scripts/upgrade [hunk 2]
```diff
@@ -14,28 +12,131 @@ source /usr/share/yunohost/helpers
...
+
+cmake . -DRUN_IN_PLACE=TRUE -DENABLE_LUAJIT=TRUE -DBUILD_CLIENT=FALSE -DBUILD_SERVER=TRUE
+
+make -j$(nproc)
+
+### Verify the checksum of a file, stored by `ynh_store_file_checksum` in the install script.
+### And create a backup of this file if the checksum is different. So the file will be backed up if the admin had modified it.
+ynh_backup_if_checksum_is_different --file="$final_path/CONFIG_FILE"
+# Recalculate and store the checksum of the file for the next upgrade.
```
Comment (kay0u) [inline scripts/upgrade]: I think you should rewrite the mintest.conf here? (maybe not)

To do so, you have to retrieve parameters like `pvp`, `damage`... and do `ynh_replace_string` like you did in install script.

File: manifest.json [hunk 3]
```diff
@@ -10,16 +10,21 @@
...
+    	"name": "Luc Didry",
+    	"email": "luc@framasoft.org",
+	"url": "https://framasoft.org"
+},
     "requirements": {
-        "yunohost": ">> 2.6.4"
+        "yunohost": ">= 3.5"
     },
-    "multi_instance": true,
+    "multi_instance": false,
```
Comment (kay0u) [inline manifest.json]: Maybe, if you replace all `mintest` by `$app` in script, you can leave this to true... But, does that make sense to have multiple instances of this game?

File: check_process [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	; Checks
+		pkg_linter=1
+		setup_sub_dir=0
+		setup_root=0
+		setup_nourl=1
+		setup_private=0
+		setup_public=0
+		upgrade=1
+		backup_restore=1
+		multi_instance=0
```
Comment (kay0u) [inline check_process]: If `multi_instance` is set to `true` in the manifest, this one have to be changed to 1.

File: README.md [hunk 5]
```diff
@@ -1,8 +1,94 @@
...
+* ARMv8-A - [![Build Status](https://ci-apps-arm.yunohost.org/ci/logs/minetest%20%28Community%29.svg)](https://ci-apps-arm.yunohost.org/ci/apps/minetest/)
+* Jessie x86-64b - [![Build Status](https://ci-stretch.nohost.me/ci/logs/minetest%20%28Community%29.svg)](https://ci-stretch.nohost.me/ci/apps/minetest/)
+
+## Limitations
+
+* This app is not compatible with LDAP.
+
+## Additional information
+
+* Migration from old package
```
Comment (kay0u) [inline README.md]: What if people didn't see this message before clicking on upgrade? You could check the version in the upgrade script, and if the version is below 5.1.0, then abort the script, and print a message saying to look at the readme.

**Comment by kay0u** [comment]
You are maintainer in this project now :-)

File: README.md [hunk 6]
```diff
@@ -1,8 +1,94 @@
-# Minetest app for YunoHost
+# Minetest for YunoHost
 
-- [Yunohost project](https://yunohost.org)
-- [Minetest website](http://www.minetest.net/)
+[![Integration level](https://dash.yunohost.org/integration/minetest.svg)](https://dash.yunohost.org/appci/app/minetest)
+
+[![Install Minetest with YunoHost](https://install-app.yunohost.org/install-with-yunohost.png)](https://install-app.yunohost.org/?app=minetest)
+
+*[Lire ce readme en français.](./README_fr.md)*
```
[the remark this answers is not in the document]
Comment (Rafi594) [reply README.md]: Fixed

File: README.md (hunk 5 above)
Comment (Rafi594) [reply README.md]: Like this ?
```
if [ "$version = 0.01" ]; then
then
ynh_die --message="You can't upgrade to this version with simple upgrade. Please read the readme for upgrading https://github.com/YunoHost-Apps/minetest_ynh#additional-information..."
fi
```

Comment (kay0u) [reply README.md]: Copy past this code inside _common.sh:
https://github.com/YunoHost-Apps/Experimental_helpers/blob/0c1738e5078e98eb3f25220a0ae9f1de94b7b04b/ynh_read_manifest/ynh_read_manifest_2

and use `ynh_app_upstream_version` to fill version, so:

```
version=ynh_app_upstream_version

if [ "$version" = "0.01" ]; then
then
ynh_die --message="You can't upgrade to this version with simple upgrade. Please read the readme for upgrading https://github.com/YunoHost-Apps/minetest_ynh#additional-information..."
fi
```

should be ok

File: check_process [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+		setup_root=0
+		setup_nourl=1
+		setup_private=0
+		setup_public=0
+		upgrade=1
+		backup_restore=1
+		multi_instance=1
+		incorrect_path=0
+		port_already_use=1
+		change_url=0
```
Comment (kay0u) [inline check_process]: Why did you disable the change_url check?

File: scripts/upgrade [hunk 8]
```diff
@@ -14,28 +12,179 @@ source /usr/share/yunohost/helpers
...
+
+make -j$(nproc)
+
+popd
+
+### Verify the checksum of a file, stored by `ynh_store_file_checksum` in the install script.
+### And create a backup of this file if the checksum is different. So the file will be backed up if the admin had modified it.
+ynh_backup_if_checksum_is_different --file="$home_path/.minetest/minetest.conf"
+# Recalculate and store the checksum of the file for the next upgrade.
+ynh_store_file_checksum --file="$home_path/.minetest/minetest.conf"
```
Comment (kay0u) [inline scripts/upgrade]: You should copy the default conf in the right place before:
`cp ../conf/minetest.conf /home/yunohost.app/$app/.minetest/`
then call all `ynh_replace_string` before `ynh_store_file_checksum`.

It should looks like that:

```
ynh_backup_if_checksum_is_different --file="$home_path/.minetest/minetest.conf"

cp ../conf/minetest.conf $home/.minetest/minetest.conf

ynh_replace_string --match_string="__APP__" --replace_string="$app" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__PORT__" --replace_string="$port" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__SERVERNAME__" --replace_string="$servername" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__DOMAIN__" --replace_string="$domain" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__ANNOUNCE__" --replace_string="$is_public" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__PVP__" --replace_string="$pvp" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__CREATIVE__" --replace_string="$creative" --target_file="$home_path/.minetest/minetest.conf"
ynh_replace_string --match_string="__DAMAGE__" --replace_string="$damage" --target_file="$home_path/.minetest/minetest.conf"

ynh_store_file_checksum --file="$home_path/.minetest/minetest.conf"
```

Comment (Rafi594) [reply scripts/upgrade]: Fixed :)

**Comment by Rafi594** [comment]
http://ci-apps-dev.yunohost.org/jenkins/job/minetest_ynh%20(Rafi59)/19/console

File: scripts/install [hunk 9]
```diff
@@ -10,115 +10,263 @@ source _common.sh
...
+yunohost service add $app --log "/var/log/$app/$app.log"
+# if using yunohost version 3.2 or more in the 'manifest.json', a description can be added
+#yunohost service add $app --description "$app daemon for XXX" --log "/var/log/$app/$app.log"
+
+#=================================================
+# RELOAD MINETEST
+#=================================================
+ynh_print_info --message="Reloading minetest..."
+
+ynh_systemd_action --action=restart --line_match="Server for gameid=minetest listening on :::30000." --log_path="/var/log/$app/minetest.log"
```
Comment (kay0u) [inline scripts/install]:
```suggestion
ynh_systemd_action --action=restart --line_match="Server for gameid="$app" listening on :::30000." --log_path="/var/log/$app/minetest.log"
```

File: scripts/restore [hunk 10]
```diff
@@ -1,99 +1,115 @@
...
-# DATA
+# RELOAD MINETEST
 #=================================================
-ynh_secure_remove "/var/games/minetest-server/"
-ynh_restore_file "/var/games/minetest-server/"
+ynh_print_info --message="Reloading minetest..."
 
-ynh_restore_file "/tmp/map.sqlite"
-echo ".restore /tmp/map.sqlite" | sqlite3 /var/games/minetest-server/.minetest/worlds/world/map.sqlite
+systemctl restart $app
```
Comment (kay0u) [inline scripts/restore]: Remove this systemctl restart $app, and ynh_systemd_action, and replace it with:

`ynh_systemd_action --action=restart --line_match="Server for gameid="$app" listening on :::30000." --log_path="/var/log/$app/minetest.log"`

`sleep 1` shouldn't be needed after that

File: scripts/upgrade [hunk 11]
```diff
@@ -14,28 +12,178 @@ source /usr/share/yunohost/helpers
...
+# Set permissions on app files
+chown -R root: $final_path
+chown -R $app: $home_path
+
+#=================================================
+# RELOAD MINETEST
+#=================================================
+ynh_print_info --message="Restarting Minetest..."
+
+systemctl restart $app
```
Comment (kay0u) [inline scripts/upgrade]:
```suggestion
ynh_systemd_action --action=restart --line_match="Server for gameid="$app" listening on :::30000." --log_path="/var/log/$app/minetest.log"
```

File: scripts/install (hunk 9 above)
Comment (kay0u) [reply scripts/install]: you can remove this `sleep 1` too

File: scripts/restore [hunk 12]
```diff
@@ -1,99 +1,115 @@
...
-is_public=$(ynh_app_setting_get $app is_public)
+domain=$(ynh_app_setting_get --app=$app --key=domain)
+final_path=$(ynh_app_setting_get --app=$app --key=final_path)
+
+#=================================================
+# CHECK IF THE APP CAN BE RESTORED
+#=================================================
+ynh_print_info --message="Validating restoration parameters..."
+
+ynh_webpath_register --app=$app --domain=$domain --path_url=/
```
Comment (kay0u) [inline scripts/restore]:
```suggestion
ynh_webpath_register --app=$app --domain=$domain --path_url="/"
```

could fix the issue

File: scripts/install [hunk 13]
```diff
@@ -10,115 +10,263 @@ source _common.sh
...
 
-codename=$(lsb_release -a 2>/dev/null | grep Codename | cut -f 2)
-test -z "$codename" && (ynh_die "codename empty")
+### If the app uses nginx as web server (written in HTML/PHP in most cases), the final path should be "/var/www/$app".
+### If the app provides an internal web server (or uses another application server such as uwsgi), the final path should be "/opt/yunohost/$app"
+final_path=/opt/yunohost/$app
+test ! -e "$final_path" || ynh_die "This path already contains a folder"
+
+# Register (book) web path
+ynh_webpath_register --app=$app --domain=$domain --path_url=/
```
Comment (kay0u) [inline scripts/install]:
```suggestion
ynh_webpath_register --app=$app --domain=$domain --path_url="/"
```

File: scripts/upgrade (hunk 11 above)
Comment (kay0u) [reply scripts/upgrade]: oups!! sorry!

We have to escape the " char:

`ynh_systemd_action --action=restart --line_match="Server for gameid=\"$app\" listening on :::30000." --log_path="/var/log/$app/minetest.log"`

is much better! (To be use everywhere)

Comment (kay0u) [reply scripts/upgrade]: Bump

**Comment by kay0u** [comment]
I fix some issue, need to be tested as well

**Comment by kay0u** [comment]
Is it normal there is no nginx conf?

**Comment by ldidry** [comment]
@kay0u Minetest is not a web service, so no nginx conf :slightly_smiling_face:

**Comment by Rafi594** [comment]
I have pushed a fix for upgrade

File: scripts/backup [hunk 14]
```diff
@@ -1,61 +1,82 @@
...
+domain=$(ynh_app_setting_get --app=$app --key=domain)
+
+#=================================================
+# STANDARD BACKUP STEPS
+#=================================================
+# BACKUP THE APP MAIN DIR
+#=================================================
+# STOP SYSTEMD SERVICE
+#=================================================
+ynh_systemd_action --service_name=$app --action="stop" --log_path="/var/log/$app/$app.log"
```
Comment (kay0u) [inline scripts/backup]: https://github.com/YunoHost-Apps/minetest_ynh/blob/4648784dfcccd50aa28aaf768890770b8aca9ef2/conf/systemd.service#L10

The log path is : `--logfile /var/log/__APP__/minetest.log`

```suggestion
ynh_systemd_action --service_name=$app --action="stop" --log_path="/var/log/$app/minetest.log"
```

File: scripts/backup [hunk 15]
```diff
@@ -1,61 +1,82 @@
...
 
-CHECK_SIZE "/etc/apt/preferences.d/00MinetestPinning"
-ynh_backup "/etc/apt/preferences.d/00MinetestPinning"
+ynh_backup --src_path="/etc/systemd/system/$app.service"
 
+#=================================================
+# END OF SCRIPT
+#=================================================
+# Restart Minetest
+ynh_systemd_action --service_name=$app --action="start" --log_path="/var/log/$app/$app.log"
```
Comment (kay0u) [inline scripts/backup]:
```suggestion
ynh_systemd_action --service_name=$app --action="start" --log_path="/var/log/$app/minetest.log"
```

File: scripts/change_url [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+ynh_replace_string --match_string="$old_domain" --replace_string="$new_domain" --target_file="/home/yunohost.app/$app/.minetest/minetest.conf"
+
+#=================================================
+# GENERIC FINALISATION
+#=================================================
+# RELOAD MINETEST
+#=================================================
+ynh_print_info --message="Reloading Minetest..."
+
+ynh_systemd_action --service_name=$app --action="restart" --log_path="/var/log/$app/$app.log"
```
Comment (kay0u) [inline scripts/change_url]:
```suggestion
ynh_systemd_action --service_name=$app --action="restart" --log_path="/var/log/$app/minetest.log"
```

File: scripts/install [hunk 17]
```diff
@@ -10,115 +10,268 @@ source _common.sh
...
+yunohost service add $app --log "/var/log/$app/minetest.log"
+# if using yunohost version 3.2 or more in the 'manifest.json', a description can be added
+#yunohost service add $app --description "$app daemon for XXX" --log "/var/log/$app/$app.log"
+
+#=================================================
+# RELOAD MINETEST
+#=================================================
+ynh_print_info --message="Reloading minetest..."
+
+ynh_systemd_action --service_name=$app --action="start" --log_path="/var/log/$app/$app.log"
```
Comment (kay0u) [inline scripts/install]:
```suggestion
ynh_systemd_action --service_name=$app --action="start" --log_path="/var/log/$app/minetest.log"
```

File: scripts/restore [hunk 18]
```diff
@@ -1,99 +1,118 @@
...
-# DATA
+# RELOAD MINETEST
 #=================================================
-ynh_secure_remove "/var/games/minetest-server/"
-ynh_restore_file "/var/games/minetest-server/"
+ynh_print_info --message="Reloading minetest..."
 
-ynh_restore_file "/tmp/map.sqlite"
-echo ".restore /tmp/map.sqlite" | sqlite3 /var/games/minetest-server/.minetest/worlds/world/map.sqlite
+ynh_systemd_action --service_name=$app --action="restart" --log_path="/var/log/$app/$app.log"
```
Comment (kay0u) [inline scripts/restore]:
```suggestion
ynh_systemd_action --service_name=$app --action="restart" --log_path="/var/log/$app/minetest.log"
```

**Comment by kay0u** [comment]
> @kay0u Minetest is not a web service, so no nginx conf slightly_smiling_face

Ok, indeed. The thing is, we have a new app in the SSO, "Minetest" after installing it, and we get the default nginx page... not really

File: scripts/backup (hunk 14 above)
Comment (Rafi594) [reply scripts/backup]: line_match seems to doesn't work 
```

Warning: yunohost.hook <lambda> - [1004.1] [WARN] The service minetest didn't fully started before the timeout.
Warning: yunohost.hook <lambda> - [1004.1] [WARN] Please find here an extract of the end of the log of the service minetest:
```

**Comment by Rafi594** [comment]
> > @kay0u Minetest is not a web service, so no nginx conf slightly_smiling_face
> 
> Ok, indeed. The thing is, we have a new app in the SSO, "Minetest" after installing it, and we get the default nginx page... not really nice for users

Maybe add a simple web page with the instructions for installing the client and connect to the server ?

**Comment by kay0u** [comment]
> Maybe add a simple web page with the instructions for installing the client and connect to the server ?

It could be great I think.

Btw, can you connect to the server after installing this app? (I can't and I don't know why, maybe because I use a ynh-dev, not accessible from the outside?..)

**Comment by Rafi594** [comment]
It works for me with a virtualbox VM but I have found a problem with minetest_game

**Comment by kay0u** [comment]
> It works for me with a virtualbox VM but I have found a problem with minetest_game
> I think it's because SOURCE_IN_SUBDIR is set to false in minetest_game.src

Yep, that's it! I let you change that

**Comment by kay0u** [comment]
https://ci-apps-dev.yunohost.org/jenkins/job/minetest_ynh%20(Kay0u)/1/console the line_match seems to work

Let wait the end of the CI before merging this: https://github.com/kay0u/minetest_ynh/tree/fix-the-line-match into you branch.

**Comment by kay0u** [comment]
After thinking about this domain story, I think we don't need to use a domain in the SSO, minetest_game doesn't really need a domain because it's not a web page, it just needs an IP with a port.
In this way, no change_url script is required

I can change the code if you want. Just, give me your point of view on this.

**Comment by Rafi594** [comment]
Yes it can also remove the unused tile in the SSO

**Comment by kay0u** [comment]
> Yes it can also remove the unused tile in the SSO

Yep that was my point. I'll propose a PR

**Comment by kay0u** [comment]
https://github.com/Rafi594/minetest_ynh/pull/3

**Comment by Rafi594** [comment]
I have tested multi_instances in a VM and the port used for the second instance is the same for the first. I think ynh_find_port doesn't check UDP

**Comment by kay0u** [comment]
> I have tested multi_instances in a VM and the port used for the second instance is the same for the first. I think ynh_find_port doesn't check UDP

Exactly, the PR to fix that: https://github.com/YunoHost/yunohost/pull/827

I fixed it on this branch: https://github.com/kay0u/minetest_ynh/tree/fix-find-port, can create a PR if needed

**Comment by Rafi594** [comment]
package_check running https://ci-apps-dev.yunohost.org/jenkins/job/minetest_ynh (Rafi59)/25/console

**Comment by kay0u** [comment]
I think it's not the right code that's being tested right now?

**Comment by Rafi594** [comment]
Yes I have strange permission denied issues in my ci-dev folder

**Comment by kay0u** [comment]
I fixed permission on this folder, you can retry now

**Comment by Rafi594** [comment]
There is strange errors with the branch fix_port_check ```
Warning: yunohost.hook <lambda> - [910.1] [WARN] The service minetest didn't fully started before the timeout.
Warning: yunohost.hook <lambda> - [910.1] [WARN] Please find here an extract of the end of the log of the service minetest:
Warning: yunohost.hook <lambda> - [910.1] -- Logs begin at Sat 2019-10-26 14:20:44 UTC, end at Sat 2019-10-26 14:34:16 UTC. --
Warning: yunohost.hook <lambda> - [910.1] Oct 26 14:30:06 pchecker_lxc systemd[1]: minetest.service: Failed to reset devices.list: Operation not permitted
Warning: yunohost.hook <lambda> - [910.1] Oct 26 14:30:06 pchecker_lxc systemd[1]: Started Minetest multiplayer server minetest.conf server config.
Error: [Main]: Error writing configuration file: "/home/yunohost.app/minetest/.minetest/worlds/world/world.mt"
```

**Comment by Rafi594** [comment]
I have found the problem, I've fixed it

**Comment by kay0u** [comment]
The restore has a problem, I'll check that latter, and I think after that we are good :-)

File: scripts/remove [hunk 19]
```diff
@@ -12,43 +12,86 @@ source /usr/share/yunohost/helpers
...
+ynh_print_info --message="Removing app main directory"
 
-# Suppression du paquet minetest
-if [ -e "/usr/lib/minetest/minetestserver" ]; then
-	echo "Remove minetest package"
-	ynh_remove_app_dependencies
+# Remove the app directory securely
+ynh_secure_remove --file="$final_path"
+ynh_secure_remove --file="$home_path"
+
```
Comment (kay0u) [inline scripts/remove]:
```suggestion
# Remove the log files
ynh_secure_remove "/var/log/$app"

```

File: scripts/restore [hunk 20]
```diff
@@ -1,99 +1,109 @@
...
-ynh_restore_file "/etc/minetest/minetest.conf"
+ynh_print_info --message="Restoring the systemd configuration..."
+
+ynh_restore_file --origin_path="/etc/systemd/system/$app.service"
+systemctl enable $app.service
+
+#=================================================
+# ADVERTISE SERVICE IN ADMIN PANEL
+#=================================================
+
```
Comment (kay0u) [inline scripts/restore]:
```suggestion

mkdir /var/log/$app -p

chown -R $app: /var/log/$app

```

**Comment by Rafi594** [comment]
[Build 
Status/badge/icon)](https://ci-apps-dev.yunohost.org/jenkins/job/minetest_ynh%20(Rafi59)/)

**Comment by kay0u** [comment]
Ready for merging I guess?

I let you the pleasure of pressing THE button

**Pull request merged**